Zend/zend_opcode.c

```diff
@@ -840,6 +839,198 @@ int pass_two_wrapper(zval *el)
 	return pass_two((zend_op_array *) Z_PTR_P(el));
 }
 
+/* The following liveliness analyzing algorithm assumes that
+ * 1) temporary variables are defined before use
+ * 2) they have linear live-ranges without "holes"
+ * 3) Opcodes never use and define the same temorary variables
+ */
+typedef struct _op_var_info {
+	struct _op_var_info *next;
+	uint32_t var;
+} op_var_info;
+
+static zend_always_inline uint32_t liveliness_kill_var(zend_op_array *op_array, zend_op *cur_op, uint32_t var, uint32_t *Tstart, op_var_info **opTs)
+{
+	uint32_t start = Tstart[var];
+	uint32_t end = cur_op - op_array->opcodes;
+	uint32_t count = 0;
+	uint32_t var_offset, j;
+
+	Tstart[var] = -1;
+	if (cur_op->opcode == ZEND_OP_DATA) {
+		end--;
+	}
+	start++;
+	if (op_array->opcodes[start].opcode == ZEND_OP_DATA
+	 || op_array->opcodes[start].opcode == ZEND_FE_FETCH_R
+	 || op_array->opcodes[start].opcode == ZEND_FE_FETCH_RW) {
```

Comment:
OP_DATA is a special opcode and it shouldn't be included into live_ranges.
FE_FETCH opcodes destroy OP1 on exception their selves.


On Tue, Jul 7, 2015 at 11:48 AM, Nikita Popov <notifications@github.com>
wrote:

> In Zend/zend_opcode.c
> <https://github.com/php/php-src/pull/1393#discussion_r34019042>:
>
> > +
> > +static zend_always_inline uint32_t liveliness_kill_var(zend_op_array *op_array, zend_op *cur_op, uint32_t var, uint32_t *Tstart, op_var_info **opTs)
> > +{
> > +	uint32_t start = Tstart[var];
> > +	uint32_t end = cur_op - op_array->opcodes;
> > +	uint32_t count = 0;
> > +	uint32_t var_offset, j;
> > +
> > +	Tstart[var] = -1;
> > +	if (cur_op->opcode == ZEND_OP_DATA) {
> > +		end--;
> > +	}
> > +	start++;
> > +	if (op_array->opcodes[start].opcode == ZEND_OP_DATA
> > +	 || op_array->opcodes[start].opcode == ZEND_FE_FETCH_R
> > +	 || op_array->opcodes[start].opcode == ZEND_FE_FETCH_RW) {
>
> Could you please explain these conditions?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/php/php-src/pull/1393/files#r34019042>.
>